### PR TITLE
feat(kotlin-sdk): expose core_wallet_next_receive_address / next_change_address (Swift parity)

### PR DESCRIPTION
**Pull request by HashEngineering**
## What

Kotlin parity for the **existing** address-derivation FFI. `rs-platform-wallet-ffi` has shipped `core_wallet_next_receive_address` / `core_wallet_next_change_address` for some time, and iOS binds the C ABI directly (`SwiftDashSDKReceiveAddressReader` → `coreWallet().nextReceiveAddress(accountIndex:)` — the DashWallet-iOS Receive screen's source). No rs-unified-sdk-jni/Kotlin plumbing existed, so Android callers (including the SDK's own KotlinExampleApp `ReceiveAddressSheet`) have been reading the Room `core_addresses` mirror instead of asking the engine.

- `rs-unified-sdk-jni/src/wallet_manager.rs`: `coreWalletNextReceiveAddress` / `coreWalletNextChangeAddress` — same `guard` / `take_pwffi_error` / `core_wallet_free_address` pattern as the neighboring core-wallet JNI calls.
- `WalletManagerNative.kt`: the two `external fun` declarations.
- `ManagedCoreWallet.kt`: public `nextReceiveAddress(accountIndex: Int = 0)` / `nextChangeAddress(accountIndex: Int = 0)`, named to mirror the Swift surface.

## Semantics (documented on the Kotlin surface)

Engine-authoritative over the in-memory used-set — "unused" means never seen on-chain. There is no issued-marker, so repeated calls return the **same** address until it receives funds (current-address semantics, matching iOS exactly, including the documented cold-start caveat: index 0 until SPV replay populates the used-set). A per-invoice `fresh_address` with an engine-side issued-marker is a separate, future ask — deliberately out of scope here.

## Why now

dash-wallet's BIP70 cutover work (dashpay/dash-wallet#1531) sources `Payment.refund_to` from the Room pool as a stopgap; once this lands in a published AAR, that Room read swaps for `coreWallet().nextReceiveAddress()` (one function body, no callers change), and the same call becomes the Receive screen's post-dashj source in Phase 2 — identical to how iOS already works.

## Verified

`cargo check -p rs-unified-sdk-jni` clean; `:sdk` Kotlin compile clean. Purely additive — no existing surface touched. Not yet exercised on-device (needs an AAR rebuild via `build_android.sh`); the FFI itself is already proven in production by iOS.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for retrieving the next unused receive address for a wallet account.
  * Added support for retrieving the next unused change address for a wallet account.
  * Account selection is supported, defaulting to the first account.
  * Invalid account indices and address retrieval errors are reported clearly through the SDK.

### CODE REVIEW

**Comment by thepastaclaw** [comment]
ℹ️ Review skipped (commit 10db4ad8)
_Last checked: 2026-08-01 22:00 UTC_

File: packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt [hunk 1]
```diff
@@ -55,6 +55,39 @@ class ManagedCoreWallet internal constructor(handle: Long) : AutoCloseable {
...
+     * The engine's next unused BIP-44 INTERNAL (change) address for
+     * [accountIndex], base58-encoded — the change-side twin of
+     * [nextReceiveAddress]; same used-set semantics and cold-start
+     * caveat. Builds pick change themselves (`setFunding`); this
+     * accessor exists for callers that must NAME a change address
+     * up front (e.g. `CoreTransactionBuilder.setChangeAddress`).
+     */
+    fun nextChangeAddress(accountIndex: Int = 0): String {
+        require(accountIndex >= 0) { "accountIndex must be non-negative, got $accountIndex" }
+        return WalletManagerNative.coreWalletNextChangeAddress(handle, accountIndex)
```
Comment (thepastaclaw) [inline packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt]: **🔴 Blocking: Map native address errors into the public SDK hierarchy**

Both new public methods call an `external fun` directly. The JNI functions use `take_pwffi_error` and `throw_sdk_exception`, so failures such as a nonexistent account or invalid native handle are raised as the internal `DashSDKException`. `DashSdkError.kt` explicitly requires every public SDK entry point that calls native code to use `mapNativeErrors`; without it, callers expecting `DashSdkError.NotFound` or `DashSdkError.PlatformWallet` receive an implementation-level exception instead. Wrap both native calls at this public boundary.

```suggestion
fun nextReceiveAddress(accountIndex: Int = 0): String {
    require(accountIndex >= 0) { "accountIndex must be non-negative, got $accountIndex" }
    return org.dashfoundation.dashsdk.errors.mapNativeErrors {
        WalletManagerNative.coreWalletNextReceiveAddress(handle, accountIndex)
    }
}

/**
 * The engine's next unused BIP-44 INTERNAL (change) address for
 * [accountIndex], base58-encoded — the change-side twin of
 * [nextReceiveAddress]; same used-set semantics and cold-start
 * caveat. Builds pick change themselves (`setFunding`); this
 * accessor exists for callers that must NAME a change address
 * up front (e.g. `CoreTransactionBuilder.setChangeAddress`).
 */
fun nextChangeAddress(accountIndex: Int = 0): String {
    require(accountIndex >= 0) { "accountIndex must be non-negative, got $accountIndex" }
    return org.dashfoundation.dashsdk.errors.mapNativeErrors {
        WalletManagerNative.coreWalletNextChangeAddress(handle, accountIndex)
    }
}
```

<sub>source: ['codex']</sub>

Comment (thepastaclaw) [inline packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt]: **🟡 Suggestion: Add device coverage for both address APIs**

No automated test exercises either new Kotlin-to-JNI path. This repository already builds the packaged native library and runs `connectedDebugAndroidTest`, while `WalletManagerRoundTripTest` creates a default wallet offline, so coverage can verify the actual exported symbols without network access. Add an instrumented test that calls both methods, confirms repeated calls return the same unused address, and confirms negative account indices are rejected; this protects the JNI signatures, handle plumbing, and documented current-address semantics.

<sub>source: ['coderabbit']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Preliminary review — Codex only

The JNI bridge's handle/index validation and C-string copy/free path are sound. The public Kotlin methods nevertheless bypass the SDK's required native-error mapping, allowing an internal exception type to escape; the class documentation is also stale, and the new JNI paths lack automated device coverage. Request changes for the public error-contract violation, with documentation and tests as non-blocking follow-ups.

Validated blockers were found in the Codex precheck. Sonnet is deferred until a fresh Codex revalidation clears the blocker gate.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — security-auditor (completed), `gpt-5.6-sol` — ffi-engineer (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet: **not run (deferred by blocker gate)**

🔴 1 blocking | 🟡 1 suggestion(s) | 💬 1 nitpick(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt`:
- [BLOCKING] packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt:73-88: Map native address errors into the public SDK hierarchy
  Both new public methods call an `external fun` directly. The JNI functions use `take_pwffi_error` and `throw_sdk_exception`, so failures such as a nonexistent account or invalid native handle are raised as the internal `DashSDKException`. `DashSdkError.kt` explicitly requires every public SDK entry point that calls native code to use `mapNativeErrors`; without it, callers expecting `DashSdkError.NotFound` or `DashSdkError.PlatformWallet` receive an implementation-level exception instead. Wrap both native calls at this public boundary.
- [SUGGESTION] packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt:73-88: Add device coverage for both address APIs
  No automated test exercises either new Kotlin-to-JNI path. This repository already builds the packaged native library and runs `connectedDebugAndroidTest`, while `WalletManagerRoundTripTest` creates a default wallet offline, so coverage can verify the actual exported symbols without network access. Add an instrumented test that calls both methods, confirms repeated calls return the same unused address, and confirms negative account indices are rejected; this protects the JNI signatures, handle plumbing, and documented current-address semantics.
```
</details>

File: packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt (hunk 1 above)
Comment (HashEngineering) [reply packages/kotlin-sdk/sdk/src/main/kotlin/org/dashfoundation/dashsdk/wallet/ManagedCoreWallet.kt]: Fixed in 70a4192164 — both accessors now wrap their native call in `mapNativeErrors`, and the stale class KDoc is refreshed. 10db4ad8ed adds host-JVM coverage: negative-index and closed-handle guards, the DashSDKException(7) → DashSdkError.NotFound mapping, and a source-scanning lint test (GateCoverageLintTest pattern) that structurally enforces the `mapNativeErrors` wrapper on both accessors so this cannot regress. Instrumented device coverage remains the documented follow-up. Ready for revalidation.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by QuantumExplorer — APPROVED** [review]
